### PR TITLE
Partially implement shell jump-to navigation

### LINKED ISSUE #14457: Unable to enter community chats if user was redirected to community via invite link
**Preconditions:** User A and User B are mutual contacts

**Steps:**

1. Create community by User A
2. Add a new channel to created community by User A
3. Send invite to community by User A to User B
4. By User B open 1-1 with User A
5. By User B tap on community link sent by User A
6. After redirection to community - join it by User B
7. Try to open existing channel by User B

**Actual result:** channel is not opened

**Expected result:** channel is opened

[geth.log](https://github.com/status-im/status-mobile/files/10114191/geth.log)
[Status.log](https://github.com/status-im/status-mobile/files/10114193/Status.log)

https://user-images.githubusercontent.com/97245802/204563233-f4c69653-5326-4162-ab12-91bf6201f93a.mp4

### Additional Information

- Status version: nightly 
- Operating System: Android, iOS

### PR DESCRIPTION
**Pull request by Parveshdhull**
## Summary
PR Implements jump-to functionality for one-to-one chat, group chat, community, and community channel.

also fixes https://github.com/status-im/status-mobile/issues/14457

### Testing Request
As PR only partially implements functionality and is still in WIP, there will be several things missing. 
Ex. for the user's last message, only text content type is supported.
So Please ignore missing stuff. And please test for any regression and any critical issue with jump-to functionality.

### Known Issue
[Scroll and back button freeze when opening chat and scrolling (iOS)](https://github.com/status-im/status-mobile/issues/14385), this issue also affecting jump-to button.

status: ready

### CODE REVIEW

**Comment by du82** [comment]
any updates to this?

**Comment by Parveshdhull** [comment]
> any updates to this?

hi @du64, Implementation is still in progress and needs a few bug fixes. And also I was hoping to extend this functionality for group chats and communities too in the same PR, so will take some time.

**Comment by du82** [comment]
Chat shows an error and card never appears

Screenshot_2022-11-27-19-07-14-97_65ad86f86fd2321a85272a99323da10f

**Comment by Parveshdhull** [comment]
> Chat shows an error and card never appears
> 
hi @du64, Thank you for testing the App. The issue is fixed, so it should not crash anymore, but switcher cards are not supported for public chats.
Currently, it only works for one-to-one, group, community screen, and community chat.

**Comment by du82** [comment]
The cards don't appear to be showing the latest sent message, or one-to-one / community icons. Is this still planned?

Screenshot_2022-11-27-19-23-45-15_65ad86f86fd2321a85272a99323da10f

**Comment by Parveshdhull** [comment]
> The cards don't appear to be showing the latest sent message, or one-to-one / community icons. Is this still planned?
> 
As mentioned in screenshot, latest sent message is still in WIP (TODO - Implement)
Also for one-to-one, community and channel icons, they are working for me. Please can you try another community which have avatar and channel thumbnails.

image

**Comment by du82** [comment]
Tested community with channel icons, it works

File: src/status_im/ui/screens/profile/user/views.cljs [hunk 1]
```diff
@@ -110,13 +110,6 @@
        :accessory           (when mnemonic
                               [components.common/counter {:size 22} 1])
        :on-press            #(re-frame/dispatch [:navigate-to :privacy-and-security])}]
-     (when config/quo-preview-enabled?
```
Comment (Parveshdhull) [inline src/status_im/ui/screens/profile/user/views.cljs:113]: Removing quo preview entry from profile. We don't use it anymore and the app just crashes on pressing it.

File: src/status_im/ui2/screens/chat/composer/view.cljs [hunk 2]
```diff
@@ -170,7 +170,7 @@
                                     (re-frame/dispatch [:dismiss-keyboard])) edit)
               [reanimated/view {:style (reanimated/apply-animations-to-style
                                         {:height shared-height}
-                                        {})}
+                                        {:z-index 2})}
```
Comment (Parveshdhull) [inline src/status_im/ui2/screens/chat/composer/view.cljs:173]: z-index 2, so that chat composer overlap floating shell button

**Comment by du82** [comment]
The jump to button covers text, and gets covered by the text field instead of being attached to it

Screenshot_2022-11-28-18-09-55-62_65ad86f86fd2321a85272a99323da10f

Screenshot_2022-11-28-18-12-50-64_65ad86f86fd2321a85272a99323da10f

**Comment by Parveshdhull** [comment]
> The jump to button covers text, and gets covered by the text field instead of being attached to it

Hi @du64, We can probably later change the behavior of the jump-to button's position with the keyboard and input box after confirmation with the design team but according to current [designs](https://www.figma.com/file/PPWkgOYlZZDxZv5SDGsZVV/Posts-%26-Attachments-for-Mobile?node-id=2272%3A149712&t=06Xh5JuSwLSAGigD-0), this (input cover jump-to) is expected behavior.

Also, if possible please use HTML syntax and smaller height for images, so they don't block PR's content.
https://stackoverflow.com/questions/24383700/resize-image-in-the-wiki-of-github-using-markdown

Thank you.

**Comment by Parveshdhull** [comment]
> The jump to button covers text

fixed second issue, thanks for finding it.

**Comment by du82** [comment]
> > The jump to button covers text, and gets covered by the text field instead of being attached to it
> 
> Hi @du64, We can probably later change the behavior of the jump-to button's position with the keyboard and input box after confirmation with the design team but according to current [designs](https://www.figma.com/file/PPWkgOYlZZDxZv5SDGsZVV/Posts-%26-Attachments-for-Mobile?node-id=2272%3A149712&t=06Xh5JuSwLSAGigD-0), this (input cover jump-to) is expected behavior.
> 
> Also, if possible please use HTML syntax and smaller height for images, so they don't block PR's content. https://stackoverflow.com/questions/24383700/resize-image-in-the-wiki-of-github-using-markdown
> 
> Thank you.

will do

File: src/status_im/ui2/screens/chat/composer/view.cljs (hunk 2 above)
Comment (J-Son89) [reply src/status_im/ui2/screens/chat/composer/view.cljs:173]: > > any updates to this?
> 
> hi @du64, Implementation is still in progress and needs a few bug fixes. And also I was hoping to extend this functionality for group chats and communities too in the same PR, so will take some time.

This p.r is already considerably large. Is it possible to add group chats and communities in follow up p.rs? 🤔

File: src/js/worklet_factory.js [hunk 3]
```diff
@@ -30,8 +30,15 @@ export function applyAnimationsToStyle(animations, style) {
   };
 };
 
-// Switcher Worklets
+// Shell Worklets
```
Comment (flexsurfer) [inline src/js/worklet_factory.js]: does it make sense to move to a separate file ? similar to bottom sheet ?

File: src/status_im2/subs/shell.cljs [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+   (sort-by :clock > (map val stacks))))
+
+(re-frame/reg-sub
+ :shell/one-to-one-chat-card
+ (fn [[_ id] _]
+   [(re-frame/subscribe [:contacts/contact-by-identity id])
+    (re-frame/subscribe [:contacts/contact-two-names-by-identity id])
+    (re-frame/subscribe [:chats/chat id])])
+ (fn [[contact names chat] [_ id]]
+   (shell/one-to-one-chat-card contact names chat id)))
```
Comment (flexsurfer) [inline src/status_im2/subs/shell.cljs]: just wondering why subscriptions function in different namespace not here?

File: src/status_im2/contexts/shell/cards/style.cljs [hunk 5]
```diff
@@ -91,11 +91,16 @@
    :style          (close-button)})
 
 (defn avatar-container []
```
Comment (flexsurfer) [inline src/status_im2/contexts/shell/cards/style.cljs]: def

File: src/status_im2/contexts/shell/constants.cljs [hunk 6]
```diff
@@ -39,3 +43,24 @@
    :chats-stack       :chats-tab-icon-opacity
    :wallet-stack      :wallet-tab-icon-opacity
    :browser-stack     :browser-tab-icon-opacity})
+
+;; TODO - Use dynamic pass-through for transparent bottom tabs,
+;; instead of hard coding value
+(def pass-through? (reagent/atom false))
```
Comment (flexsurfer) [inline src/status_im2/contexts/shell/constants.cljs]: doesn't look like constant ?

File: src/status_im2/contexts/shell/cards/view.cljs [hunk 7]
```diff
@@ -62,16 +65,29 @@
...
+                                :override-theme :dark}]
+
+    constants/community-card
+    (if (:source avatar-params)
+      [fast-image/fast-image
+       {:source (:source avatar-params)
+        :style  (style/community-avatar customization-color)}]
+      ;; TODO - Update to fall back community avatar once designs are available
+      [rn/view {:style (style/community-avatar customization-color)}
+       [text/text {:weight :semi-bold
```
Comment (flexsurfer) [inline src/status_im2/contexts/shell/cards/view.cljs]: quo/text

File: src/status_im2/contexts/shell/cards/view.cljs [hunk 8]
```diff
@@ -62,16 +65,29 @@
...
-                                             :style  (style/community-avatar)}]))
+    constants/one-to-one-chat-card
+    [user-avatar/user-avatar
+     (merge {:ring?             false
+             :size              :medium
+             :status-indicator? false}
+            avatar-params)]
+
+    constants/private-group-chat-card
+    [group-avatar/group-avatar {:color          customization-color
```
Comment (flexsurfer) [inline src/status_im2/contexts/shell/cards/view.cljs]: quo/group-avatar

File: src/status_im2/contexts/shell/cards/view.cljs [hunk 9]
```diff
@@ -62,16 +65,29 @@
...
-                              :size              :medium
-                              :status-indicator? false}
-                             avatar-params)]
-    :group-messaging [group-avatar/group-avatar {:color          customization-color
-                                                 :size           :large
-                                                 :override-theme :dark}]
-    :community-card  [fast-image/fast-image {:source (:source avatar-params)
-                                             :style  (style/community-avatar)}]))
+    constants/one-to-one-chat-card
+    [user-avatar/user-avatar
```
Comment (flexsurfer) [inline src/status_im2/contexts/shell/cards/view.cljs]: quo/user-avatar

File: src/status_im2/contexts/shell/cards/view.cljs [hunk 10]
```diff
@@ -1,20 +1,23 @@
...
+            [clojure.string :as string]
             [quo2.foundations.colors :as colors]
+            [react-native.fast-image :as fast-image]
             [quo2.components.markdown.text :as text]
             [quo2.components.buttons.button :as button]
             [quo2.components.counter.counter :as counter]
+            [status-im2.contexts.shell.cards.style :as style]
             [quo2.components.tags.status-tags :as status-tags]
+            [status-im2.contexts.shell.constants :as constants]
             [quo2.components.avatars.user-avatar :as user-avatar]
```
Comment (flexsurfer) [inline src/status_im2/contexts/shell/cards/view.cljs]: please replace all references by using one quo reference

File: src/status_im2/contexts/shell/view.cljs [hunk 11]
```diff
@@ -38,38 +41,77 @@
...
-                            :style {:margin-top (:top insets)}}]
-      [placeholder]
-      [rn/scroll-view {:style {:padding-horizontal 20
-                               :flex-direction     :row}}
-       [quo/text {:size   :heading-1
-                  :weight :semi-bold
-                  :style  {:color      colors/white
-                           :margin-top 12}}
-        (i18n/label :t/jump-to)]]])])
+  (let [switcher-cards @(re-frame/subscribe [:shell/sorted-switcher-cards])
```
Comment (flexsurfer) [inline src/status_im2/contexts/shell/view.cljs]: rf/sub

File: src/status_im2/contexts/shell/view.cljs [hunk 12]
```diff
@@ -38,38 +41,77 @@
...
+        shell-margin   (/ (- width 320) 3)] ;; 320 - two cards width
+    [safe-area/consumer
+     (fn [insets]
+       [rn/view {:style {:top              0
+                         :left             0
+                         :right            0
+                         :bottom           -1
+                         :position         :absolute
+                         :background-color colors/neutral-100}}
+        [common.home/top-nav {:type  :shell
```
Comment (flexsurfer) [inline src/status_im2/contexts/shell/view.cljs:99]: quo/top-nav

File: src/status_im2/contexts/shell/core.cljs [hunk 13]
```diff
@@ -0,0 +1,151 @@
+(ns status-im2.contexts.shell.core
```
Comment (flexsurfer) [inline src/status_im2/contexts/shell/core.cljs]: status-im2.contexts.shell.events

File: src/status_im2/contexts/shell/core.cljs [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+  (:require [utils.re-frame :as rf]
+            [re-frame.core :as re-frame]
+            [status-im.constants :as status-constants]
+            [status-im2.navigation.events :as navigation]
+            [status-im.react-native.resources :as resources]
+            [status-im2.contexts.shell.animation :as animation]
+            [status-im2.contexts.shell.constants :as constants]))
```
Comment (flexsurfer) [inline src/status_im2/contexts/shell/core.cljs]: i would say it's better to have `constants` and `shell.constants`

**Review by flexsurfer — COMMENTED** [review]
great job!

**Review by flexsurfer — APPROVED** [review]

File: src/status_im2/contexts/shell/view.cljs (hunk 12 above)
Comment (Parveshdhull) [reply src/status_im2/contexts/shell/view.cljs:99]: we don't have top-nav as quo component anymore (https://github.com/status-im/status-mobile/issues/14408)

File: src/status_im/ui2/screens/chat/composer/view.cljs (hunk 2 above)
Comment (ibrkhalil) [reply src/status_im/ui2/screens/chat/composer/view.cljs:173]: Is this an iOS only thing, Should we add `elevation` ?

Comment (Parveshdhull) [reply src/status_im/ui2/screens/chat/composer/view.cljs:173]: https://reactnative.dev/docs/layout-props#zindex
> zIndex controls which components display on top of others.

Its not for elevation or shadow, it only controls order of view in document tree

**Comment by Parveshdhull** [comment]
Hi @pavloburykh, Thank you for finding https://github.com/status-im/status-mobile/issues/14457.

This PR migrates chat to new navigation, so I guess it will be better to implement the fix directly for the new navigation.

So, now when the user will click on community (from the message), then opened chat will be closed and the community will be open, and on pressing back user will return to the communities screen instead of chats.

(Once animations are implemented, transitions like this (between screens), will be much smoother 🤞)

**Comment by du82** [comment]
The scroll arrow should match the height of the jump button.

Screenshot_2022-11-29-14-58-47-23_65ad86f86fd2321a85272a99323da10f

**Comment by Parveshdhull** [comment]
> The scroll arrow should match the height of the jump button.

We need to remove the old down arrow button and use the one provided by the floating shell button.
Will do this later in a separate PR. Thank you

**Comment by pavloburykh** [comment]
> #14457

Great! Thank you @Parveshdhull

**Review by jakubgs — APPROVED** [review]
Approving only Nix change.

**Comment by qoqobolo** [comment]
Hey @Parveshdhull , thanks for working on this feature! 
It looks good to me, although testing is still in progress 🙂
However, there is one issue regarding the fix for https://github.com/status-im/status-mobile/issues/14457

### ISSUE 1: `Not found` screen with the `Fetch community` button when following an invitation link on iOS

**Steps:**
1. Device1 adds Device2 (iOS) to contacts
2. Device1 creates a community and sends invitation via Community >  `+ button` > `Invite people` to Device2
3. Device2 opens 1-1 chat with Device1 and follows the link

I can't reproduce this behavior in develop (also, I was able to reproduce the original issue only for Android devices).
Can you check please?

https://user-images.githubusercontent.com/67952253/204807065-873a4e04-ffa9-49c2-8691-7ef5a240fcc0.MP4

**Comment by qoqobolo** [comment]
### ISSUE 2/Q: Should different channels of the same community be saved as different cards on the `Jump to` screen?

**Steps:**
1. Create/join a community with at least 2 channels
2. Open channel1 > tap `Jump to`
3. Get back to the community
4. Open channel2 > tap `Jump to`

**Actual behavior:** channel1 is replaced with channel2
**Expected (?) behavior:** the card for channel1 keeps displaying, a new card for channel2 is created (as it working for group chats, for example)

https://user-images.githubusercontent.com/67952253/204812711-05da7244-14d3-4335-bc07-6e519b5abfdb.mp4

My doubts about whether this is within the scope of this PR is based on the description of the feature 🙂

> PR Implements jump-to functionality for one-to-one chat, group chat, community, and **community channel.**

Also, adding some designs

https://www.figma.com/file/eDfxTa9IoaCMUy5cLTp0ys/Shell-for-Mobile?node-id=0%3A1&t=AiAoFDQsBEveN3i5-0

Screenshot 2022-11-30 at 14 40 33

### ISSUE 3/Q: Should the Jump to screen be cleared after reopening/relogin?


https://user-images.githubusercontent.com/67952253/204819620-962acace-d786-434a-b4ac-443b10c7b785.mp4

**Comment by Parveshdhull** [comment]
hi @qoqobolo, Thank you for testing the PR and finding these issues.

1. will fix it, IOS always needs special attention :) 
2. One community card represents, one application. And the content of that card represents the user's location when minimized.
So, there will be always only one card per community.
In design these are two cards because they are different communities (Dribble and SuperRare), only avatars are the same (probably by accident)
3. Yes for now we are not persisting shell cards, so they will be cleared at login. But we are going to work on this later, so the app state remains the same across sessions.

**Comment by qoqobolo** [comment]
> In design these are two cards because they are different communities (Dribble and SuperRare), only avatars are the same (probably by accident)

OMG, sorry I should have been more attentive, the avatars got me confused :(

Thanks for your answers!

**Comment by Parveshdhull** [comment]
Issue 1 should be fixed now

**Comment by qoqobolo** [comment]
Confirming that Issue 1 is fixed, thank you @Parveshdhull!
No other critical issues were found, so PR can be merged.

**Pull request merged**